_posts/2021-02-10-hindsight-on-2020.md

```diff
@@ -0,0 +1,165 @@
+`jpeg-decoder` crate that has fixed quite a lot of outstanding bugs in the
+decoding process; The newly added bindings for `avif` decoders and encoders;
+And not to forget improved and copyless image operations.
+
+There are still many outstanding issues and unimplemented improvements. From
+the point of view of a maintainer, some of the most pressing ones are listed in
+the next sections, sorted roughly by topic.
+
+## Fallbacks and Error logs
+
+The decoders are strict, in many ways. There exist a few fallbacks here and
+there, but not enough. The justification for this is that being too lenient in
+accepted formats leads to an ossification of formats where particular
+non-standard behavior by the most popular encoders is de-facto standardized
+without appearing in any specification. For many older formats that ship has,
+however, long sailed. We should thus accept the popular interpretation (e.g.
+ImageMagick) of such extensions.
+
+The issues goes a bit farther than this. It would be great to have a mode where
+such details and nit-picks are remarked upon but decoding continues. Meanwhile
+we do not want to dump them to `stderr`.
+
+Related issues:
+* <https://github.com/image-rs/image-png/issues/193>: Chunk buffer size is inconsistent with std recommendation
+* <https://github.com/image-rs/image-png/issues/254>: Don't verify adler32 checksum by default 
+* <https://github.com/image-rs/jpeg-decoder/issues/90>: Decoding Twice Error
+* <https://github.com/image-rs/jpeg-decoder/issues/169>: Corrupted JPEG does not result in visible error code
+
+## Supplementary information
+
+The `image` library currently does not expose a consistent interface to
+supplementary information such as color spaces, comments, copyright,
+orientation, and other parts of EXIF. This starts at concrete treatment of
+extension chunks in the decoder libraries and goes all the way to design work
+for integrating those in the `ImageDecoder` trait and the `Reader`.
+
+* <https://github.com/image-rs/image-gif/issues/40>: Allow reading loop control extension (so that it can be written again)
+* <https://github.com/image-rs/image-gif/issues/79>: The treatment of extensions
+* <https://github.com/image-rs/image-png/issues/116>: Support for encoding metadata chunks
+* <https://github.com/image-rs/image-png/issues/221>: Treatment of unknown chunks
+* <https://github.com/image-rs/image/issues/258>: Need more portable image coordinates.
+* <https://github.com/image-rs/image/issues/1371>: Orientation field not found after load image and save
+
+## Limits during decoding
+
+Loading an image from a remote server, only to find out that it decompresses to
+several gigabytes of memory and totally grinds your system to a halt/crashes
+your program isn't great. We'd like to ensure that the decoders all have memory
+and/or runtime limits that they check and abide to. This also helps with
+fuzzing as it turns allows controlling for such use explicitly. Note that most
```

Comment:
nit: turns -> in turn

Reply:
0x570D401AD99fF3f33CDdEdCcb4cfD29D3CDA33BB